File: 234-palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {

        // STEP 1:- BREAK FROM THE MIDDLE 

        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // STEP 2:- reverse the second half od the LL

        ListNode rightHead = reverse (slow) ;


        // Step 3:- Compare for a palindrome 

        ListNode temp = head ;
        ListNode temp2 = rightHead ;
        while(temp2 != null){
            if(temp.val != temp2.val){
                return false ;
            }
            temp = temp.next ;
            temp2 = temp2.next ;
        }

        return true ;
    }

    private ListNode reverse(ListNode head){
        ListNode temp = head ;
        ListNode last = null ;

        while(head != null){
            temp = head ;
            head = head.next ;
            temp.next = last ;
            last = temp ;
        }

        return temp ;
    }


}

File: 234-palindrome-linked-list/palindrome-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {

    
    // 1. break from middle
    slow := head 
    fast := head
    // var prev ListNode

    for fast != nil && fast.Next != nil{
        // prev = slow
        slow = slow.Next
        fast = fast.Next.Next
    }

    // 2. reverse from the middle 

    head2 := reverse(slow)

    // 3. comapre
    i := head
    j := head2

    for i != head2 && j != nil{
        if i.Val != j.Val{
            return false 
        }
        i = i.Next
        j = j.Next
    }

    return true
    
}

func reverse(head *ListNode) *ListNode{
    var prev *ListNode 

    for head != nil {
        curr := head 
        head = head.Next
        curr.Next = prev
        prev = curr
    }

    return prev
}